3d assets blog post (#1357)
huggingface/blog: 3d-assets.md

---
title: "Practical 3D Asset Generation: A Step-by-Step Guide"
thumbnail: /blog/assets/124_ml-for-games/thumbnail-3d.png
authors:
- user: dylanebert
---

> <h1>Practical 3D Asset Generation: A Step-by-Step Guide</h1>

Reviewer: Should we include `AI` or `ML` in the title?

<!-- {blog_metadata} -->
<!-- {authors} -->

## Introduction

> Generative AI is an instrumental part of artistic workflows. However, as detailed in my [earlier post](https://huggingface.co/blog/ml-for-games-3), text-to-3D lags behind 2D in terms of practical applicability. This is beginning to change. Today, we'll be revisiting practical workflows for 3D Asset Generation and taking a step-by-step look at how to integrate Generate AI in a PS1-style 3D workflow.

Reviewer: ```suggestion
Generative AI has become an instrumental part of artistic workflows for game development. However, as detailed in my [earlier post](https://huggingface.co/blog/ml-for-games-3), text-to-3D lags behind 2D in terms of practical applicability. This is beginning to change. Today, we'll be revisiting practical workflows for 3D Asset Generation and taking a step-by-step look at how to integrate Generative AI in a PS1-style 3D workflow.
```

> <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/124_ml-for-games/3d/result.png" alt="final result"/>

Reviewer: Just a reminder to run the images through an image optimizer, I got ~20% reduction on this one :)

Also, unless you want to tweak the presentation style (see [here](https://github.com/huggingface/blog/blob/c07213c4250a6dd6233cb1c0b28f09db38e7dd8b/vision_language_pretraining.md?plain=1#L85) for image captions) it's probably ok to use the markdown syntax.

Why the PS1 style? Because it's much more forgiving to the low fidelity of current text-to-3D models, and allows us to go from text to usable 3D asset with as little effort as possible.

## Step 1: Generate a 3D Model

> Start by visiting the Shap-E Hugging Face Space [here](https://huggingface.co/spaces/hysts/Shap-E) or down below. This space uses the open-source [Shap-E model](https://github.com/openai/shap-e) from OpenAI to generate 3D models from text.

Reviewer: ```suggestion
Start by visiting the Shap-E Hugging Face Space [here](https://huggingface.co/spaces/hysts/Shap-E) or down below. This space uses the open-source [Shap-E model](https://github.com/openai/shap-e), a recent diffusion model from OpenAI to generate 3D models from text.
```

(note to self: we should migrate the Space to diffusers if we can)

<gradio-app theme_mode="light" space="hysts/Shap-E"></gradio-app>

> Enter "Dilapidated Shack" as your prompt and click 'Generate'. When you're happy with the model, download it for the next step.

Reviewer: ```suggestion
Enter "Dilapidated Shack" as your prompt and click 'Generate'. When you're happy with the generated 3D model, download it for the next step.
```

(just to prevent minor confusion with the AI model)

<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/124_ml-for-games/3d/shape.png" alt="shap-e space"/>

## Step 2: Import and Decimate the Model

Next, open [Blender](https://www.blender.org/download/) (version 3.1 or higher). Go to File -> Import -> GLTF 2.0, and import your downloaded file. You may notice that the model has way more polygons than recommended for many practical applications, like games.

> <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/124_ml-for-games/3d/import.png" alt="importing the model in blender"/>

Reviewer: In this case the text is a bit blurry. Can we maybe zoom to the menus if it's not too much work?

To reduce the polygon count, select your model, navigate to Modifiers, and choose the "Decimate" modifier. Adjust the ratio to a low number (i.e. 0.02). This is probably *not* going to look very good. However, in this tutorial, we're going to embrace the low fidelity.

## Step 3: Install Dream Textures

> To add textures to our model, we'll be using [Dream Textures](https://github.com/carson-katri/dream-textures). Follow the instructions on the [official repository](https://github.com/carson-katri/dream-textures) to download and install the addon.

Reviewer: ```suggestion
To add textures to our model, we'll be using [Dream Textures](https://github.com/carson-katri/dream-textures), a stable diffusion texture generator for Blender. Follow the instructions on the [official repository](https://github.com/carson-katri/dream-textures) to download and install the addon.
```

<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/124_ml-for-games/3d/dreamtextures.png" alt="installing dream textures"/>

> Once installed and enabled, open the addon preferences. Search for and download the [texture-diffusion](https://huggingface.co/dream-textures/texture-diffusion) model.

Reviewer: Maybe cross-link to https://huggingface.co/blog/ml-for-games-4 for an overview of 2D asset / texture generation?

<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/124_ml-for-games/3d/generate.png" alt="generating a texture"/>

To apply the texture, select your model and navigate to 'Material'. Add a new material, and under 'base color', click the dot and choose 'Image Texture'. Finally, select your newly generated texture.

## Step 5: UV Mapping

Time for UV mapping, which wraps our 2D texture around the 3D model. Select your model and press 'Tab' to enter Edit Mode. Then, press 'U' to unwrap the model and choose 'Smart UV Project'.

To preview your textured model, switch to rendered view (hold 'Z' and select 'Rendered'). You can scale up the UV map to have it tile seamlessly over the model. Remember that we're aiming for a retro PS1 style, so don't make it too nice.

<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/124_ml-for-games/3d/uv.png" alt="uv mapping"/>

## Step 6: Export the Model

When you're happy with your model, it's time to export it. Navigate to File -> Export -> FBX, and voila! You have a usable 3D Asset.

## Step 7: Import in Unity

> Finally, let's see our model in action. Import it in [Unity](https://unity.com/download) (or your game engine of choice). To recreate a nostalgic PS1 aesthetic, I've customized it with custom vertex-lit shading, no shadows, lots of fog, and glitchy post-processing. You can read more about recreating the PS1 aesthetic [here](https://www.david-colson.com/2021/11/30/ps1-style-renderer.html).

Reviewer: ```suggestion
Finally, let's see our model in action. Import it in [Unity](https://unity.com/download) or your game engine of choice. To recreate a nostalgic PS1 aesthetic, I've customized it with custom vertex-lit shading, no shadows, lots of fog, and glitchy post-processing. You can read more about recreating the PS1 aesthetic [here](https://www.david-colson.com/2021/11/30/ps1-style-renderer.html).
```
